ibis/expr/deferred.py

```diff
@@ -0,0 +1,121 @@
+        return self.__class__(
+            lambda expr, func=func, self=self, args=args, kwargs=kwargs: func(
+                self.resolve(expr),
+                *(_resolve(arg, expr=expr) for arg in args),
+                **{
+                    name: _resolve(arg, expr=expr)
+                    for name, arg in kwargs.items()
+                },
+            )
+        )
+
+    def __getattr__(self, name: str) -> Deferred:
+        return self._wrap(getattr, name)
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Deferred:
+        return self._wrap(
+            lambda expr, *args, **kwargs: expr(*args, **kwargs),
+            *args,
+            **kwargs,
+        )
+
+    def __add__(self, other: Any) -> Deferred:
+        return self._wrap(operator.add, other)
+
+    def __radd__(self, other: Any) -> Deferred:
+        return self._wrap(operator.add, other)
+
+    def __sub__(self, other: Any) -> Deferred:
+        return self._wrap(operator.sub, other)
+
+    def __rsub__(self, other: Any) -> Deferred:
+        return self._wrap(operator.add, other)
+
+    def __mul__(self, other: Any) -> Deferred:
+        return self._wrap(operator.mul, other)
+
+    def __rmul__(self, other: Any) -> Deferred:
+        return self._wrap(operator.mul, other)
+
+    def __truediv__(self, other: Any) -> Deferred:
+        return self._wrap(operator.truediv, other)
+
+    def __rtruediv__(self, other: Any) -> Deferred:
+        return self._wrap(operator.truediv, other)
+
+    def __floordiv__(self, other: Any) -> Deferred:
+        return self._wrap(operator.floordiv, other)
+
+    def __rfloordiv__(self, other: Any) -> Deferred:
+        return self._wrap(operator.floordiv, other)
+
+    def __pow__(self, other: Any) -> Deferred:
+        return self._wrap(operator.pow, other)
+
+    def __rpow__(self, other: Any) -> Deferred:
+        return self._wrap(operator.pow, other)
+
+    def __mod__(self, other: Any) -> Deferred:
+        return self._wrap(operator.mod, other)
+
+    def __rmod__(self, other: Any) -> Deferred:
+        return self._wrap(operator.mod, other)
+
+    def __eq__(self, other: Any) -> Deferred:  # type: ignore
+        return self._wrap(operator.eq, other)
+
+    def __ne__(self, other: Any) -> Deferred:  # type: ignore
+        return self._wrap(operator.ne, other)
+
+    def __lt__(self, other: Any) -> Deferred:
+        return self._wrap(operator.lt, other)
+
+    def __le__(self, other: Any) -> Deferred:
+        return self._wrap(operator.le, other)
+
+    def __gt__(self, other: Any) -> Deferred:
+        return self._wrap(operator.gt, other)
+
+    def __ge__(self, other: Any) -> Deferred:
+        return self._wrap(operator.ge, other)
+
+    def __invert__(self) -> Deferred:
+        return self._wrap(operator.invert)
+
+    def __neg__(self) -> Deferred:
+        return self._wrap(operator.neg)
+
+
+T = TypeVar("T")
+
+
+@functools.singledispatch
+def _resolve(arg: T, **_: Any) -> T:
```

Comment:
Is this for making it pluggable? If not could we avoid using singledispatch?

Reply:
Removed.